crypto/evp_extra/evp_extra_test.cc

```diff
@@ -3072,12 +3093,287 @@ TEST(EVPExtraTest, ParamgenCallbacks) {
   EXPECT_TRUE(app_data.state);
 
   for (int i = 0; i < keygen_info; i++) {
-    // DH_generate_parameters_ex makes a final call to `BN_GENCB_call(cb, 3, 0)`
     if(i == 0) {
-      EXPECT_EQ(EVP_PKEY_CTX_get_keygen_info(ctx.get(), i), 3);
+      EXPECT_EQ(EVP_PKEY_CTX_get_keygen_info(ctx.get(), i), GetParam().keygen_info_0);
     } else {
-      EXPECT_EQ(EVP_PKEY_CTX_get_keygen_info(ctx.get(), i), 0);
+      EXPECT_EQ(EVP_PKEY_CTX_get_keygen_info(ctx.get(), i), GetParam().keygen_into_1);
+    }
+  }
+}
+
+static bssl::UniquePtr<EVP_PKEY> dsa_paramgen(int nbits, const EVP_MD* md, bool copy) {
+  bssl::UniquePtr<EVP_PKEY> params(nullptr);
+  EVP_PKEY* pkey_raw = NULL;
+
+  auto maybe_copy = [&](bssl::UniquePtr<EVP_PKEY_CTX> *ctx) -> bool {
+    if (copy) {
+      ctx->reset(EVP_PKEY_CTX_dup(ctx->get()));
+    }
+    return *ctx != nullptr;
+  };
+
+  // Construct a EVP_PKEY_CTX
+  bssl::UniquePtr<EVP_PKEY_CTX> ctx(EVP_PKEY_CTX_new_id(EVP_PKEY_DSA, nullptr));
+  if(!ctx) {
+    goto dsa_paramgen_end;
+  }
+  if(!maybe_copy(&ctx)) {
+    goto dsa_paramgen_end;
+  }
+  // Initialize for paramgen
+  if(1 != EVP_PKEY_paramgen_init(ctx.get())) {
+    goto dsa_paramgen_end;
+  }
+  // Set the prime length
+  if(1 != EVP_PKEY_CTX_set_dsa_paramgen_bits(ctx.get(), nbits)) {
+    goto dsa_paramgen_end;
+  }
+  // Set the generator
+  if(1 != EVP_PKEY_CTX_set_dsa_paramgen_md(ctx.get(), md)) {
+    goto dsa_paramgen_end;
+  }
+  if(1 != EVP_PKEY_paramgen(ctx.get(), &pkey_raw)) {
+    goto dsa_paramgen_end;
+  }
+  params.reset(pkey_raw);
+dsa_paramgen_end:
+  return params;
```

Comment:
Nit: Can we consolidate these into one if statement and just return `params` if any fail? Same for the other helper functions below as well.

Our C++ code has mostly gotten rid of `goto` statements in favor of using scoped pointers. I don't think there's a necessity to use them here.